### ISSUE TITLE
Should expression templates be enabled by default as of 2024?

### ISSUE THREAD

**Issue by marcovc** [boostorg/multiprecision#622]
Hi,

I just spent  a couple of days hunting a bug that was caused by using "auto" with this library. I then found out that this is expected:

![image](https://github.com/boostorg/multiprecision/assets/624308/12dd47cb-c60c-4c0d-a431-5a83b65a70ab)
 
I wonder if it is a correct decision today, where "auto" is now very well established into the language, to ship a library that by default computes complete wrong results when using it with "auto".

**Comment by ckormanyos** [comment]
> spent a couple of days hunting a bug that was caused by using `auto` with this library.

Yes this is a known _feature_. You raise a really good question.
- Enable `ExpressionTemplates` by default (i.e., leave the library alone)?
- Disable `ExpressionTemplates` by default? Do we have issues with a breaking changeß
- Eliminate `ExpressionTemplates` entirely? Wow lots of typing and do we have issues with a breaking change?

The final, most obtrusive, option being one I'd preferably not engage on.

I wonder what John thinks of this. We worked hard to get those big numbers. Your intuition might be right from today`s perspective. Maybe one day down the road a compiler switch that actively enables _legacy_ could do the trick to get `auto` working from this point forward...

I wonder what @jzmaddock thinks of such evolutions. I'm kind of a _leave_ _it_ _alone_ and documetn it guy.

**Comment by jzmaddock** [comment]
+1 on leave it alone, rationale:

1) It's a pretty well known issue, there is a proposal to fix this: https://isocpp.org/files/papers/p0672r0.pdf which uses expression templates as it's motivating factor.  The issue even effects proxy objects like `std::vector<bool>::operator[]`, so we are hardly alone here.
2) I know we have people that rely on the expression templates for performance reasons - for some of the science applications which can take _days_ to run, even a few percent speedup is vital.
3) It is documented, and while I realise it's fairly well down the page, there's a big "caution" in the intro that mentions this: https://www.boost.org/doc/libs/1_85_0/libs/multiprecision/doc/html/boost_multiprecision/intro.html.  Of course if you're anything like me, you may well skip the docs entirely if you can ;)

**Comment by jzmaddock** [comment]
I've pinged Joel to see if there's any news on the issue (or an alternative fix in C++17/20/23).

**Comment by piotr5** [comment]
I'm coming from the ultimate++ library which is very similar to boost. there they intensively use the "pick-semantics" meaning that no container ever has a copy-constructor and instead the default is to use the move-constructor. and that move-constructor actually does move things, meaning that whatever was owned by a container is then moved to the new container's ownership just because of applying std::move to it. so the object itself remains in the memory it occupied and only the pointer is swapped with the new container's pointer.  on the downside, this only works if both containers understand that pick-semantics, but in case of this warning's actual example: maybe just add some expression-building functions for R-value input? what that function should do is to just disown the cpp_int storing whatever temporaty int and give it the live-time of the expression-template instead. i.e. allow expressions to store the limbs temporarily but do it in constant time without copying. example-code:

```
#include <boost/multiprecision/cpp_int.hpp>
#include <iostream>

using R=boost::multiprecision::cpp_int;

template<class T>
auto TT(const T& a, T c){
    const T t={a[0],c[1]-a[1]};
    return a[0]*std::move(t[1]);
}
int main(){
    std::array<R,2> d{7,2},e{5,11},f{17,19};
    std::cout << TT(d,f)<<"\n";
}
```

this should not result in bad_alloc for compilers being strict about temporary objects. instead after std::move above the value stored in the 2nd entry of the array t should be in an undefined state while the expression object returned should contain the temporary produced on the way. hard to express the pick-semantics in the documentation though...

**Comment by ckormanyos** [comment]
I have thought a long time about this issue.

No is my (subjective) answer to the original query.

Compilers have become better, yes indeed. But changing all the docs and templates is _not_ adding significant value at the expense of huge development time - which is scant (at least for the moment).

So I would suggest to leave the setup the way it is. And then close this issue accordingly.

Cc: @jzmaddock and @mborland
